### PR TITLE
[EAGLE-3552][EAGLE-3567] - Add inference params and multimodal-embedder model

### PR DESCRIPTION
**Pull request by phatvo9**
[EAGLE-3552] - Add inference parameters:
* Modify model.py and inference.py to receives them
* Generate JSON file for uploading
* Test with inference parameters
* update doc

[EAGLE-3552] - add multimodal-embedder model type

[EAGLE-3552]: https://clarifai.atlassian.net/browse/EAGLE-3552?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EAGLE-3552]: https://clarifai.atlassian.net/browse/EAGLE-3552?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: clarifai/models/model_serving/docs/inference_parameters.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+## Inference paramters
```
Comment (HarmitMinhas96) [inline clarifai/models/model_serving/docs/inference_parameters.md:1]: Few instances in this file:
*paramters -> parameters

similarly in clarifai/models/model_serving/models/test.py

File: clarifai/models/model_serving/models/model_types.py [hunk 2]
```diff
@@ -234,3 +234,32 @@ def parse_predictions(self, input_data: np.ndarray):
     return inference_response
 
   return parse_predictions
+
+
+def multimodal_embedder(func: Callable):
+  """
+  Visual embedder type output parser.
+  Generates embeddings for an input image.
```
Comment (HarmitMinhas96) [inline clarifai/models/model_serving/models/model_types.py:242]: This should mention this supports both image and/or text

File: clarifai/models/model_serving/model_config/inference_parameter.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
+import json
+from dataclasses import asdict, dataclass, field
+from typing import Any, List
+
+
+@dataclass(frozen=True)
+class InferParamType:
+  BOOL: int = 1
+  STRING: int = 2
+  NUMBER: int = 3
```
Comment (HarmitMinhas96) [inline clarifai/models/model_serving/model_config/inference_parameter.py:10]: Very soon we're going to have `ENCRYPTED_STRING = 21;` support which will be needed here. I think it's ok to get ahead and add that now. That would be used for things like API Keys where we never want to return it via the API. The user would provide a string value. I see the code currently depends on their only being one string field type, so maybe we can always default to string, but have a way for users to specify it's an encrypted string too.
cc @wemoveon2

File: clarifai/models/model_serving/model_config/inference_parameter.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    return list_params
+
+  def export(self, path: str):
+    list_params = self.get_list_params()
+    with open(path, "w") as fp:
+      json.dump(list_params, fp, indent=2)
+
+  def validate(self, **kwargs) -> dict:
+    output_kwargs = {k: v.default_value for k, v in self._dict_params.items()}
+    assert not (kwargs != {} and self.params == []), "kwargs are rejected since `params` is empty"
```
Comment (HarmitMinhas96) [inline clarifai/models/model_serving/model_config/inference_parameter.py:86]: Might be simpler written:
```kwargs == {} or self.params != []```

File: clarifai/models/model_serving/docs/inference_parameters.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+When making prediction, you may need to change some paramters to adjust the result. Those paramters will be passed through `paramters()` of a request in triton python model.
+
+In order to send it to `**kwargs` of `get_predictions` in `inference.py`, there are 2 ways:
+1. You can send any arbitrary parameters via clarifai API.
```
Comment (HarmitMinhas96) [inline clarifai/models/model_serving/docs/inference_parameters.md:6]: We're removing this, they will need to be defined by the model owner

File: clarifai/models/model_serving/docs/inference_parameters.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+2. You can define some parameters and they will be visible and adjustable on Clarifai model views.
```
Comment (HarmitMinhas96) [inline clarifai/models/model_serving/docs/inference_parameters.md:7]: *visible

**Review by HarmitMinhas96 — COMMENTED** [review]
Could you possibly split the multi-modal embedder into it's own smaller PR?

File: clarifai/models/model_serving/model_config/inference_parameter.py (hunk 3 above)
Comment (wemoveon2) [reply clarifai/models/model_serving/model_config/inference_parameter.py:10]: yup add `ENCRYPTED_STRING` here and i'll make the changes to BE to support the field type.

File: clarifai/models/model_serving/models/default_test.py [hunk 7]
```diff
@@ -30,58 +32,73 @@ class DefaultTestInferenceModel(unittest.TestCase):
...
+    _kwargs = self.inference_parameters.validate(**kwargs)
+    return inspect.unwrap(self.triton_python_model.inference_obj.get_predictions)(
+        self.triton_python_model.inference_obj, input_data, **_kwargs)
 
-  def _get_preprocess(self):
+  def _get_preprocess(self, input):
     """ preprocess if input is image """
-    if "image" in self.triton_model_input_name:
-      h, w, _ = self.triton_model_config.input[0].dims
+    if "name" in input.name:
```
Comment (HarmitMinhas96) [inline clarifai/models/model_serving/models/default_test.py:51]: I think "name" is supposed to remain "image" here?

File: clarifai/models/model_serving/docs/inference_parameters.md [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+When making prediction, you may need to change some paramaters to adjust the result. Those paramaters will be passed through `paramaters()` of a request in triton python model.
+
+In order to send it to `**kwargs` of `get_predictions` in `inference.py`, you can define some parameters and they will be visible and adjustable on Clarifai model view.
+
+This document helps you to create your inference parameters that can be visibale and adjustable easily on Clarifai platform. The defined parameters will be sent as `json` file when you use `clarifai-upload-model` cli.
+
+### JSON file structure:
+The file contains a list of object has 4 fields:
+* `path` (str): name of your parameter, it must be valid as python variable
+* `field_type` (int): the parameter data type is one of {1,2,21,3}, it means {boolean, string, encrypted_string, number} respectively. `Number` means `int` or `float`. `Encrypted_string` is simple a string can be used to store your secrets like API key...
```
Comment (HarmitMinhas96) [inline clarifai/models/model_serving/docs/inference_parameters.md:12]: "Encrypted_string is simple a string can be used to store your secrets like API key..."
->
"Encrypted_string is a string that can be used to store your secrets, like API key. The API will not return the values for this as plaintext."

**Review by HarmitMinhas96 — APPROVED** [review]

File: clarifai/models/model_serving/docs/inference_parameters.md (hunk 8 above)
Comment (phatvo9) [reply clarifai/models/model_serving/docs/inference_parameters.md]: thanks for your suggestion

File: clarifai/models/model_serving/docs/inference_parameters.md (hunk 5 above)
Comment (phatvo9) [reply clarifai/models/model_serving/docs/inference_parameters.md]: Sure, error will arise if user uses undefined parameters in test cases

**Review by HarmitMinhas96 — APPROVED** [review]

**Pull request merged**